File: internal/bluetooth/reconnector.go
package bluetooth

import (
	"context"
	"log"
	"os"
	"strconv"
	"time"
)

// StartReconnectLoop periodically attempts to connect to paired+trusted devices not connected
func StartReconnectLoop(ctx context.Context, mgr BluetoothManagerInterface, interval time.Duration) {
	go func() {
		log.Printf("Reconnect: loop started (interval=%s)", interval)
		ticker := time.NewTicker(interval)
		defer ticker.Stop()

		for {
			select {
			case <-ctx.Done():
				log.Printf("Reconnect: loop stopped")
				return
			case <-ticker.C:
				adapters, err := mgr.GetAdapters()
				if err != nil {
					log.Printf("Reconnect: get adapters error: %v", err)
					continue
				}
				for _, a := range adapters {
					devices, err := mgr.GetDevices(a.Path)
					if err != nil {
						log.Printf("Reconnect: get devices error for %s: %v", a.Address, err)
						continue
					}
					for _, d := range devices {
						if d.Paired && d.Trusted && !d.Connected {
							// Attempt connection with shorter timeout inside ConnectDevice
							if err := mgr.ConnectDevice(a.Path, d.Address); err != nil {
								log.Printf("Reconnect: connect %s failed: %v", d.Address, err)
							} else {
								log.Printf("Reconnect: connect attempt for %s initiated", d.Address)
							}
						}
					}
				}
			}
		}
	}()
}

// GetReconnectInterval returns interval from env RECONNECT_INTERVAL_SECONDS or default seconds
func GetReconnectInterval(defaultSeconds int) time.Duration {
	if v := os.Getenv("RECONNECT_INTERVAL_SECONDS"); v != "" {
		if n, err := strconv.Atoi(v); err == nil && n > 0 {
			return time.Duration(n) * time.Second
		}
	}
	return time.Duration(defaultSeconds) * time.Second
}

File: src/bluetooth/reconnector.rs
use std::sync::Arc;
use std::time::Duration;

use super::BluetoothManager;

/// Spawns a background task that periodically reconnects paired+trusted devices
/// that are not currently connected.
pub fn start_reconnect_loop(manager: Arc<BluetoothManager>, interval: Duration) {
    tokio::spawn(async move {
        tracing::info!("Reconnect: loop started (interval={:?})", interval);
        let mut ticker = tokio::time::interval(interval);

        loop {
            ticker.tick().await;

            let adapters = match manager.get_adapters().await {
                Ok(a) => a,
                Err(e) => {
                    tracing::warn!("Reconnect: get adapters error: {}", e);
                    continue;
                }
            };

            for adapter in &adapters {
                let devices = match manager.get_devices(&adapter.path).await {
                    Ok(d) => d,
                    Err(e) => {
                        tracing::warn!(
                            "Reconnect: get devices error for {}: {}",
                            adapter.address,
                            e
                        );
                        continue;
                    }
                };

                for device in &devices {
                    if device.paired && device.trusted && !device.connected {
                        match manager.connect_device(&adapter.path, &device.address).await {
                            Ok(_) => tracing::info!(
                                "Reconnect: connect attempt for {} initiated",
                                device.address
                            ),
                            Err(e) => tracing::warn!(
                                "Reconnect: connect {} failed: {}",
                                device.address,
                                e
                            ),
                        }
                    }
                }
            }
        }
    });
}

/// Returns the reconnect interval from `RECONNECT_INTERVAL_SECONDS` env var
/// or the provided default (in seconds).
pub fn get_reconnect_interval(default_seconds: u64) -> Duration {
    std::env::var("RECONNECT_INTERVAL_SECONDS")
        .ok()
        .and_then(|v| v.parse::<u64>().ok())
        .filter(|&n| n > 0)
        .map(Duration::from_secs)
        .unwrap_or(Duration::from_secs(default_seconds))
}
